### PR TITLE
Change process spawning to inherit the parent's signal mask by default

### PR DESCRIPTION
**Pull request by sunshowers**
Previously, the signal mask was always reset when a child process is
started. This breaks tools like `nohup` which expect `SIGHUP` to be
blocked for all transitive processes.

With this change, the default behavior changes to inherit the signal mask.

This also changes the signal disposition for `SIGPIPE` to only be changed if the `#[unix_sigpipe]` attribute isn't set.

### CODE REVIEW

**Comment by sunshowers** [comment]
(I don't know of a way to represent the stack cleanly in GitHub, unlike in Differential or other tools.)

**Comment by thomcc** [comment]
Ideally it would be cleanly separated from the other (I know this is a bit of a pain). That would make it easier to FCP and land on its own.

**Comment by sunshowers** [comment]
It's not really possible to cleanly separate. Best one could do is to change the ordering so that #100737 depends on this rather than the other way round.

To avoid duplicated work I would be happy to do that after the FCP regarding this is approved.

**Comment by thomcc** [comment]
Sure. I'll assign to someone who can start an FCP for this (I really need to get to https://github.com/rust-lang/rfcbot-rs/issues/306... maybe on Monday...). For their context, this is based on this comment from @zackw: https://github.com/rust-lang/rust/pull/100737#issuecomment-1222672934

r? @joshtriplett

**Comment by sunshowers** [comment]
Is there anything blocking this? @joshtriplett @thomcc

Once this is accepted I'd be happy to change the ordering of the PRs and resolve merge conflicts, etc.

File: library/std/src/os/unix/process.rs [hunk 1]
```diff
@@ -179,6 +179,103 @@ pub trait CommandExt: Sealed {
...
+    /// ```no_run
+    /// #![feature(process_sigmask)]
+    /// #
+    /// use std::process::Command;
+    /// use std::os::unix::process::CommandExt;
+    ///
+    /// Command::new("sleep")
+    ///     .arg("10")
+    ///     // On most platforms, SIGINT is signal 2.
+    ///     .block_signal(2)?
```
Comment (joshtriplett) [inline library/std/src/os/unix/process.rs:216]: The comment should say something like "prefer libc::SIGINT if available"

**Comment by joshtriplett** [comment]
I think the handling for SIGPIPE needs some updates, for two reasons:

- This should take https://github.com/rust-lang/rust/pull/97802 into account, rather than unconditionally unblocking.
- If the user uses block_signal on SIGPIPE, we should respect that and not override it.

**Comment by Enselic** [comment]
> This should take [Support `#[unix_sigpipe = "inherit|sig_dfl"]` on `fn main()` to prevent ignoring `SIGPIPE` #97802](https://github.com/rust-lang/rust/pull/97802) into account, rather than unconditionally unblocking.

Would you mind elaborating a bit on that please? In my mind that attribute only affects what the runtime does to SIGPIPE before fn main() is invoked. I don't think it should affect the spawning of child processes? (Except indirectly)

**Comment by joshtriplett** [comment]
"except indirectly" is exactly my concern here. Right now, and with this patch, Rust resets SIGPIPE to default in the child process. I think if the user sets the sigpipe disposition with that attribute, we should allow that to be inherited unless overridden.

**Comment by Enselic** [comment]
Hmm, in that case we should probably make `#[unix_sigpipe = "..."]` a module attribute rather than an attribute on `fn main()`, because otherwise the behaviour of inheritance of parent signal masks can't be configured for library code. But if we make it an attribute that can be used with libraries, how do we handle when two different libraries that are in use choose different values on the attribute?

To me, these problems suggest that the `#[unix_sigpipe = "..."` "belongs" to `fn main()`, since without an `fn main()`, no code is generated that can configure `SIGPIPE` before `fn main()` is invoked.

I think it is useful to think about what we would do if the Rust runtime would never have started to touch `SIGPIPE` in 2014. If that would have been the case, how would we have written the code in this PR? And I think the answer to that is "we do not treat `SIGPIPE` any different than any other signal", so that seems to me to be what we should do also today here.

But I am of course open to be convinced otherwise.

**Comment by joshtriplett** [comment]
@Enselic I agree, and I wish we could do that, but I think that would break backwards compatibility at the moment.

I'm proposing that if something sets the attribute *at all*, we treat that as an opt in to not touching the signal *here*.

**Comment by sunshowers** [comment]
@joshtriplett Some questions:

1.

How should `reset_sigpipe`: https://github.com/rust-lang/rust/blob/8c6ce6b91b172f77c795a74bfeaf74b865146b3f/library/std/src/sys/unix/mod.rs#L156-L181

    communicate with process spawning code? My first thought was to use a static `OnceLock`, but I'm wondering if it'll run into issues given how early in the process this happens. Do you have other/better suggestions?

2. Note that this PR touches `setsigmask` (AIUI a mask of signals currently blocked from delivery, though can be delivered in the future), not `setsigdefault` (signals that are reset to default signal handling, but which can still potentially be temporarily blocked by my understanding). If `block_signal` is called, that doesn't directly affect `setsigdefault`, only `setsigmask`.
  a. How should the behavior of `setsigdefault` change under `block_signal`?
  b. If `SIGPIPE` is currently *blocked* in this process, and `unblock_signal` is called, what should happen to the behavior of `setsigdefault`?

3. What's the best way to test this behavior? It seems like there are a number of different combinations involved here.

**Comment by sunshowers** [comment]
Changes since the last version:

1. Made it so that this is now free-standing and no longer dependent on #100737.
2. Per @joshtriplett's command, changed the code so that `SIGPIPE` is no longer set if `#[unix_sigpipe]` is specified.
3. Added documentation to the unix-sigpipe unstable feature about 2, noting in particular that there's now a difference between not specifying it and specifying `"sig_ign"`.

The `block_signal` considerations can be deferred to #100737 -- they're no longer relevant with this version.

I still don't know how to test the `SIGPIPE` change yet, unfortunately.

File: library/std/src/os/unix/process.rs (hunk 1 above)
Comment (sunshowers) [reply library/std/src/os/unix/process.rs]: (This is no longer relevant as of this change -- I'll fix #100737 up with it.)

File: library/std/src/sys/unix/process/process_unix.rs [hunk 2]
```diff
@@ -326,30 +326,26 @@ impl Command {
...
+            // If #[unix_sigpipe] is not specified, reset SIGPIPE to SIG_DFL for backward compatibility.
+            //
+            // #[unix_sigpipe] is an opportunity to change the default here.
+            if !unix_sigpipe_attr_specified() {
+                #[cfg(target_os = "android")] // see issue #88585
+                {
+                    let mut action: libc::sigaction = mem::zeroed();
+                    action.sa_sigaction = libc::SIG_DFL;
+                    cvt(libc::sigaction(libc::SIGPIPE, &action, ptr::null_mut()))?;
+                }
```
Comment (sunshowers) [inline library/std/src/sys/unix/process/process_unix.rs:342]: Is the behavior on Android correct? I'm not sure.

**Comment by sunshowers** [comment]
OK so @joshtriplett @Enselic I've spent a couple of hours on this and I still don't know how to write an integration test for this using compiletest.

The thing I would like to test is: ensure that the process spawning code doesn't touch the `SIGPIPE` signal handler if `#[unix_sigpipe]` is specified. The closest example I've found is https://github.com/rust-lang/rust/blob/4a14677239bbd9a3892d165ee24513c1a05820e4/src/test/ui/process/process-sigpipe.rs#L22-L36. So I tried writing a variant of this test that sets `#[unix_sigpipe = "sig_ign"]` etc.

How do I actually test the behavior? Well, some options I can think of:
1. **Use a shell script.** How? The `yes | head` has a pretty clear error case if we want to test `SIG_DFL`. However, if we want to test `SIG_IGN`, this just produces `yes: standard output: Broken pipe` on my machine, which is text I don't think we can rely on to be stable.
2. **Use a Rust binary.** This would be ideal. But the only examples I can find execute the current binary, which is clearly not suitable since it'll be influenced by the `#[unix_sigpipe]` directive. We'd want to have a Rust test execute *another* Rust binary, and I don't know if compiletest supports that.
3. **Use Python?** I don't know if this is a feasible option.

As a meta comment, I'm still not sure what this has to do with my change, which is to inherit the signal *mask*, and is unrelated to signal *handlers*. I think this should be a separate PR that is unrelated to this.

**Comment by joshtriplett** [comment]
The current approach looks good to me, and I think it'd be reasonable to work on a test separate from this PR.

File: compiler/rustc_session/src/config/sigpipe.rs [hunk 3]
```diff
@@ -16,7 +16,10 @@ pub const SIG_IGN: u8 = 2;
...
 
-/// `SIG_IGN` has been the Rust default since 2014. See
+/// `DEFAULT` is a separate value here, but is resolved to `SIG_IGN` in
+/// `library/std/src/sys/unix/mod.rs`.
+///
+/// Note that `SIG_IGN` has been the Rust default since 2014. See
 /// <https://github.com/rust-lang/rust/issues/62569>.
 #[allow(dead_code)]
-pub const DEFAULT: u8 = SIG_IGN;
+pub const DEFAULT: u8 = 0;
```
Comment (Enselic) [inline compiler/rustc_session/src/config/sigpipe.rs:25]: I agree `0` is a good value for `DEFAULT`, but can we perhaps move it to the top of the file so that the constants becomes sequential? Or maybe change the value to `4`, but that seems less attractive to me.

We should also update

https://github.com/rust-lang/rust/blob/bb5a01617589b5e3ece5a36435fc285bfd13c9a4/library/std/src/rt.rs#L92

to not say "3 values", but instead say "4 values". Or perhaps even better: rephrase it to not specify an exact count.

File: library/std/src/sys/unix/mod.rs [hunk 4]
```diff
@@ -166,21 +167,45 @@ pub unsafe fn init(argc: isize, argv: *const *const u8, sigpipe: u8) {
                 pub const INHERIT: u8 = 1;
                 pub const SIG_IGN: u8 = 2;
                 pub const SIG_DFL: u8 = 3;
+                pub const DEFAULT: u8 = 0;
```
Comment (Enselic) [inline library/std/src/sys/unix/mod.rs:170]: (Also move this to the top if you move the definition to the top)

File: library/std/src/sys/unix/mod.rs [hunk 5]
```diff
@@ -166,21 +167,45 @@ pub unsafe fn init(argc: isize, argv: *const *const u8, sigpipe: u8) {
...
 
-            let handler = match sigpipe {
-                sigpipe::INHERIT => None,
-                sigpipe::SIG_IGN => Some(libc::SIG_IGN),
-                sigpipe::SIG_DFL => Some(libc::SIG_DFL),
+            let (sigpipe_attr_specified, handler) = match sigpipe {
+                sigpipe::INHERIT => (true, None),
+                sigpipe::SIG_IGN => (true, Some(libc::SIG_IGN)),
+                sigpipe::SIG_DFL => (true, Some(libc::SIG_DFL)),
+                sigpipe::DEFAULT => (false, Some(libc::SIG_IGN)),
```
Comment (Enselic) [inline library/std/src/sys/unix/mod.rs:177]: (Also move this to the top if you move the definition to the top)

File: src/doc/unstable-book/src/language-features/unix-sigpipe.md [hunk 6]
```diff
@@ -36,7 +36,7 @@ hello world
 
 Set the `SIGPIPE` handler to `SIG_IGN` before invoking `fn main()`. This will result in `ErrorKind::BrokenPipe` errors if you program tries to write to a closed pipe. This is normally what you want if you for example write socket servers, socket clients, or pipe peers.
 
-This is what libstd has done by default since 2014. Omitting `#[unix_sigpipe = "..."]` is the same as having `#[unix_sigpipe = "sig_ign"]`.
+This is what libstd has done by default since 2014. (However, see the note on child processes below). Omitting `#[unix_sigpipe = "..."]` is the same as having `#[unix_sigpipe = "sig_ign"]`.
```
Comment (Enselic) [inline src/doc/unstable-book/src/language-features/unix-sigpipe.md:39]: We should remove the phrase

> Omitting `#[unix_sigpipe = "..."]` is the same as having `#[unix_sigpipe = "sig_ign"]`.

now, right?

Comment (sunshowers) [reply src/doc/unstable-book/src/language-features/unix-sigpipe.md]: Yep, done!

File: compiler/rustc_session/src/config/sigpipe.rs (hunk 3 above)
Comment (sunshowers) [reply compiler/rustc_session/src/config/sigpipe.rs]: Done, thanks.

File: library/std/src/sys/unix/mod.rs (hunk 5 above)
Comment (sunshowers) [reply library/std/src/sys/unix/mod.rs]: Done.

**Comment by sunshowers** [comment]
OK, review comments addressed. I think this is ready for FCP and to land :)

**Comment by joshtriplett** [comment]
This looks good to me. Since this changes user-visible behavior, and *can't* be put behind a nightly feature gate, I think we need to FCP this.

@rfcbot merge

**Comment by Xiretza** [comment]
Failed in rollup: https://github.com/rust-lang/rust/pull/103247#issuecomment-1284310551

**Comment by sunshowers** [comment]
Thanks. I've addressed the failures -- I think this is ready to land again.

**Comment by thomcc** [comment]
FYI if you want to run the CI checks without waiting for bors to get around to trying it, see https://rustc-dev-guide.rust-lang.org/tests/ci.html?highlight=ci#using-ci-to-test for a guide. (Just be sure not to land any CI changes, of course)

**Comment by sunshowers** [comment]
Thanks. Apparently this last comment was just from my previous land attempt. Hopefully this one goes smoother.

**Comment by passcod** [comment]
Just some feedback that this [broke cargo-watch](https://github.com/watchexec/cargo-watch/issues/249) as we set sigmask on the parent process and relied on the sigmask reset for children to run them as the user expected.

**Comment by sunshowers** [comment]
You can fix it locally by temporarily blocking the signals in the spawning thread. See https://github.com/nextest-rs/nextest/blob/2505a731eadd7b7bc7355a53e753a05d822a358d/nextest-runner/src/double_spawn.rs#L167 for an example.

**Comment by passcod** [comment]
It just seems weird because the previous thing was deliberately done: https://github.com/rust-lang/rust/pull/25784
